modyn/trainer_server/internal/trainer/remote_downsamplers/remote_rs2_downsampling.py

```diff
@@ -0,0 +1,81 @@
+import logging
+import random
+from typing import Any, Optional
+
+import torch
+from modyn.trainer_server.internal.trainer.remote_downsamplers.abstract_remote_downsampling_strategy import (
+    AbstractRemoteDownsamplingStrategy,
+)
+
+logger = logging.getLogger(__name__)
+
+
+class RemoteRS2Downsampling(AbstractRemoteDownsamplingStrategy):
+    """
+    Method adapted from REPEATED RANDOM SAMPLING FOR MINIMIZING THE TIME-TO-ACCURACY OF LEARNING (Okanovic+, 2024)
+    https://openreview.net/pdf?id=JnRStoIuTe
+    """
+
+    def __init__(
+        self,
+        pipeline_id: int,
+        trigger_id: int,
+        batch_size: int,
+        params_from_selector: dict,
+        per_sample_loss: Any,
+        device: str,
+    ) -> None:
+        super().__init__(pipeline_id, trigger_id, batch_size, params_from_selector, device)
+        self.forward_required = False
+        self.supports_bts = False
+        self._all_sample_ids: list[int] = []
+        self._subsets: list[list[int]] = []
+        self._current_subset = -1
+        self._with_replacement: bool = params_from_selector["replacement"]
+        self._max_subset = -1
+        self._first_epoch = True
+
+    def init_downsampler(self) -> None:
+        pass  # We take care of that in inform_samples
+
+    def inform_samples(
+        self,
+        sample_ids: list[int],
+        forward_output: torch.Tensor,
+        target: torch.Tensor,
+        embedding: Optional[torch.Tensor] = None,
+    ) -> None:
+        # We only need to collect the sample information once
+        if self._first_epoch:
+            self._all_sample_ids.extend(sample_ids)
+
+    def _epoch_step_wr(self, target_size: int) -> None:
+        self._subsets = [self._all_sample_ids[:target_size]]
+        self._current_subset = 0
+
+    def _epoch_step_r(self, target_size: int) -> None:
+        self._max_subset = len(self._all_sample_ids) // target_size
+        self._current_subset += 1
+        if self._current_subset >= self._max_subset or len(self._subsets) == 0:
```

Comment:
In the very first round, this will be the case. We need some form of flag to indicate reshuffling, either after we passed through all subsets or have not done anything yet

Reply:
Thanks! that explains it